### DISCUSSION TITLE
How can I create a relation between different Schemas?

### DISCUSSION THREAD

**Discussion by SelfDevTV** [pkosiec/mongo-seeding#157] [Q&A] [answered]
Hi,

first of all awesome library!

How can I create a relation between different collections?

**Example:**

In the example folder we have two collections: `categories` and `posts`. 

What if I go into the posts and do something like this:

```
const posts = [
  {
    id: getObjectId('post1'),
    title: 'Lorem ipsum',
    description: 'Sample Post 1 description',
    tags: ['sample', 'tags'],
    author: {
      name: 'Author',
      email: 'test@example.com',
    },
    categories: [ // the specified category id's from 1-categories/categories.js],
    comments: [],
    creationDate: new Date(),
  }
]
```

And another example with the same data. 
What if I also want to have something like this in my category (one to many it's called I think):

    {    _id: // the id of the category that's clear,
        name: "Additional Category,
        post: // the corresponding posts object id or the post itself
    }

Is this achievable and I missed it somehow in the docs?

**Comment by pkosiec** [comment]
Hello @SelfDevTV!
Thank you for very kind words.

Actually, in the first example from `examples` there is already a relation defined. MongoDB is a document database, so the relations work a little bit different that the ones in SQL databases.

```
const posts = [
  {
    categories: [ // the specified category id's from 1-categories/categories.js], // <- here is relation defined; specifically - one-to-many. Category can have multiple posts
  }
]
```

In the second example you provided, in the `post` field there should be an array with [post IDs](https://docs.mongodb.com/manual/tutorial/model-referenced-one-to-many-relationships-between-documents/), or, alternatively, you can [embed the posts directly](https://docs.mongodb.com/manual/tutorial/model-embedded-one-to-one-relationships-between-documents/). 

I would recommend reading the official MongoDB docs about document relations - they are very solid:
- https://docs.mongodb.com/manual/tutorial/model-embedded-one-to-one-relationships-between-documents/
- https://docs.mongodb.com/manual/tutorial/model-embedded-one-to-many-relationships-between-documents/
- https://docs.mongodb.com/manual/tutorial/model-referenced-one-to-many-relationships-between-documents/

Does it answer your question?

Best regards,
Pawel

**Comment by SelfDevTV** [comment]
Hi @pkosiec 

Thank you for the fast reply.

Where in the first example can I find this relation? 
Because I added the additional prop `categories` in my example above. In the examples I can't find any relation to `categories`.

Maybe you can link me to this specific example you mean where this is set up correctly. 

The second example:

You are right, since a post can have many categories a Category can have many posts. But still how can I relate one to another? How do I access the id's of the posts? Can't find it. 

Thank you very much,
Dom

**Comment by pkosiec** [comment, marked as answer]
Oh, right - sorry, my bad. In the `examples` directory there are no relations indeed. I thought you've just copied it from the repo, but you actually modified it. And you did it right - this is how the relation should look like, if you want to do it by document references.

So, it should look like this:

Posts:
```javascript
const posts = [
  {
    id: getObjectId('post1'),
    title: 'Lorem ipsum',
    description: 'Sample Post 1 description',
    tags: ['sample', 'tags'],
    author: {
      name: 'Author',
      email: 'test@example.com',
    },
    categories: [getObjectId("sample"), getObjectId("random")],
    comments: [],
    creationDate: new Date(),
  }
]

module.exports = posts;
```

Categories:
```javascript
module.exports = [
{
  id: getObjectId("sample"),
  name: "Sample category",
},
{
  id: getObjectId("random"),
  name: "Random category",
},
{
  id: getObjectId("another-one"),
  name: "Another category",
},
];
```

All IDs are predictable, that's why you can specify that your post is assigned to "random" and "sample" categories.

I hope now it's clear how it is done. The whole idea is to have predictable IDs, and that's it.

And here's what's the `getObjectId` helper does:
```javascript
const getObjectId = name => {
  const hash = createHash('sha1')
    .update(name, 'utf8')
    .digest('hex');

  return new ObjectId(hash.substring(0, 24));
};
```

So, it creates a hash, and makes an ObjectId instance of part of the hash. We know that the result will be always the same, if the input name is equal.

**Comment by SelfDevTV** [comment]
Thank you very much. That clears it up for me :) 
I get it now. 

I just thought that this helper always returns a random hash. But when the name is the same the id is the same, now I get it :) 

Thank you and have a nice day.
Greets from Austria

**Comment by pkosiec** [comment]
You're welcome! Thanks for pointing it out - I agree that this example should contain one relation for clarity. I will update it for 3.2.1. 

Greetings from Poland 🙂

**Comment by SelfDevTV** [comment]
Tried this out and it works like a charm. Now this lib has everything I need for my projects purposes. Thank you very much again.

**Comment by pkosiec** [comment]
I'm glad you like it 🙂 Enjoy using it!
